File: web/src/services/feishu.ts
import request from './api'
import type { ApiResponse } from '../types'

export interface SendCardRequest {
  receive_id: string
  receive_id_type: 'open_id' | 'user_id' | 'union_id' | 'email' | 'chat_id'
  card_data: {
    title?: string
    services: Array<{
      name: string
      object_id: string
      actions: string[]
      branches?: string[]
    }>
  }
}

export interface SendMessageRequest {
  receive_id: string
  receive_id_type: 'open_id' | 'user_id' | 'union_id' | 'email' | 'chat_id'
  msg_type: 'text' | 'post' | 'image' | 'interactive'
  content: string
}

export interface FeishuApp {
  id?: number
  name: string
  app_id: string
  app_secret: string
  webhook?: string
  project: string
  description?: string
  status: string
  is_default: boolean
  created_at?: string
  updated_at?: string
}

export interface FeishuBot {
  id?: number
  name: string
  webhook_url: string
  project?: string
  secret?: string
  description?: string
  status: string
  message_template_id?: number
  created_at?: string
  updated_at?: string
}

export const feishuApi = {
  sendCard: (data: SendCardRequest): Promise<ApiResponse> => {
    return request.post('/feishu/api/send-card', data)
  },

  sendMessage: (data: SendMessageRequest): Promise<ApiResponse> => {
    return request.post('/feishu/send-message', data)
  },

  getVersion: (): Promise<ApiResponse<{ version: string }>> => {
    return request.get('/feishu/version')
  }
}

// 飞书应用管理
export interface AppBindings {
  jenkins_instances: Array<{ id: number; name: string; url: string }>
  k8s_clusters: Array<{ id: number; name: string }>
}

export const feishuAppApi = {
  list: (page = 1, pageSize = 100): Promise<ApiResponse<{ list: FeishuApp[]; total: number }>> => {
    return request.get('/feishu/app', { params: { page, page_size: pageSize } })
  },

  get: (id: number): Promise<ApiResponse<FeishuApp>> => {
    return request.get(`/feishu/app/${id}`)
  },

  create: (data: Partial<FeishuApp>): Promise<ApiResponse<FeishuApp>> => {
    return request.post('/feishu/app', data)
  },

  update: (id: number, data: Partial<FeishuApp>): Promise<ApiResponse<FeishuApp>> => {
    return request.put(`/feishu/app/${id}`, data)
  },

  delete: (id: number): Promise<ApiResponse> => {
    return request.delete(`/feishu/app/${id}`)
  },

  setDefault: (id: number): Promise<ApiResponse> => {
    return request.post(`/feishu/app/${id}/default`)
  },

  getBindings: (id: number): Promise<ApiResponse<AppBindings>> => {
    return request.get(`/feishu/app/${id}/bindings`)
  }
}

// 飞书机器人管理
export const feishuBotApi = {
  list: (page = 1, pageSize = 100): Promise<ApiResponse<{ list: FeishuBot[]; total: number }>> => {
    return request.get('/feishu/bot', { params: { page, page_size: pageSize } })
  },

  get: (id: number): Promise<ApiResponse<FeishuBot>> => {
    return request.get(`/feishu/bot/${id}`)
  },

  create: (data: Partial<FeishuBot>): Promise<ApiResponse<FeishuBot>> => {
    return request.post('/feishu/bot', data)
  },

  update: (id: number, data: Partial<FeishuBot>): Promise<ApiResponse<FeishuBot>> => {
    return request.put(`/feishu/bot/${id}`, data)
  },

  delete: (id: number): Promise<ApiResponse> => {
    return request.delete(`/feishu/bot/${id}`)
  }
}

// 飞书回调管理
export const feishuCallbackApi = {
  getStatus: (): Promise<ApiResponse<{ running_apps: string[] }>> => {
    return request.get('/feishu/callback/status')
  },

  refresh: (): Promise<ApiResponse<{ running_apps: string[] }>> => {
    return request.post('/feishu/callback/refresh')
  }
}

// 飞书消息日志
export interface FeishuMessageLog {
  id: number
  created_at: string
  msg_type: string
  receive_id: string
  receive_id_type: string
  content: string
  title?: string
  source: string
  status: string
  error_msg?: string
  app_id?: number
}

export const feishuLogApi = {
  list: (page = 1, pageSize = 20, msgType = '', source = ''): Promise<ApiResponse<{ list: FeishuMessageLog[]; total: number }>> => {
    return request.get('/feishu/logs', { params: { page, page_size: pageSize, msg_type: msgType, source } })
  },

  get: (id: number): Promise<ApiResponse<FeishuMessageLog>> => {
    return request.get(`/feishu/logs/${id}`)
  }
}

// 用户搜索
export interface FeishuUser {
  user_id?: string
  open_id?: string
  union_id?: string
  name?: string
  en_name?: string
  email?: string
  mobile?: string
  avatar?: {
    avatar_72?: string
    avatar_240?: string
    avatar_640?: string
    avatar_origin?: string
  }
  department_ids?: string[]
}

export const feishuUserApi = {
  search: (query: string): Promise<ApiResponse<FeishuUser[]>> => {
    return request.post('/feishu/user/search', { query })
  },

  get: (id: string, userIdType = 'user_id'): Promise<ApiResponse<FeishuUser>> => {
    return request.get(`/feishu/user/${id}`, { params: { user_id_type: userIdType } })
  },

  setToken: (appId: string, userToken: string, refreshToken: string): Promise<ApiResponse> => {
    return request.post('/feishu/user/token', { app_id: appId, user_token: userToken, refresh_token: refreshToken })
  },

  getTokenStatus: (appId?: string): Promise<ApiResponse<{ has_token: boolean; refresh_token: string }>> => {
    return request.get('/feishu/user/token/status', { params: { app_id: appId } })
  },

  getAuthorizeUrl: (): string => {
    return '/app/api/v1/feishu/oauth/authorize'
  }
}

// 群聊管理
export interface FeishuChat {
  chat_id: string
  name: string
  description?: string
  avatar?: string
  owner_id?: string
  owner_id_type?: string
  chat_mode?: string
  chat_type?: string
  external?: boolean
  tenant_key?: string
}

export const feishuChatApi = {
  list: (pageToken = '', pageSize = 20): Promise<ApiResponse<{ list: FeishuChat[]; page_token: string }>> => {
    return request.get('/feishu/chat', { params: { page_token: pageToken, page_size: pageSize } })
  },

  create: (name: string, description: string, userIds: string[], userIdType = 'user_id'): Promise<ApiResponse<{ chat_id: string }>> => {
    return request.post('/feishu/chat', { name, description, user_ids: userIds, user_id_type: userIdType })
  },

  addMembers: (chatId: string, userIds: string[], userIdType = 'user_id'): Promise<ApiResponse> => {
    return request.post(`/feishu/chat/${chatId}/members`, { user_ids: userIds, user_id_type: userIdType })
  }
}

File: internal/models/notification/feishu.go
// Package notification 定义消息通知相关的数据模型
// 本文件包含飞书相关的模型定义
package notification

import (
	"time"

	"gorm.io/gorm"
)

// ==================== 飞书模型 ====================

// FeishuBot 飞书机器人模型
// 存储飞书 Webhook 机器人配置
type FeishuBot struct {
	ID                uint      `gorm:"primarykey" json:"id"`
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
	Name              string    `gorm:"size:100;not null" json:"name"`                   // 机器人名称
	WebhookURL        string    `gorm:"size:500;not null" json:"webhook_url"`            // Webhook URL
	Project           string    `gorm:"size:100;default:''" json:"project"`              // 关联项目
	Secret            string    `gorm:"size:100" json:"secret"`                          // 签名密钥
	Description       string    `gorm:"type:text" json:"description"`                    // 描述
	Status            string    `gorm:"size:20;default:'active';not null" json:"status"` // 状态
	MessageTemplateID *uint     `gorm:"index" json:"message_template_id"`                // 消息模板ID
	CreatedBy         *uint     `gorm:"index" json:"created_by"`
}

// TableName 指定表名
func (FeishuBot) TableName() string {
	return "feishu_bots"
}

// FeishuApp 飞书应用模型
// 存储飞书开放平台应用配置
type FeishuApp struct {
	ID          uint      `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	Name        string    `gorm:"size:100;not null" json:"name"`                         // 应用名称
	AppID       string    `gorm:"column:app_id;size:100;not null" json:"app_id"`         // 飞书 App ID
	AppSecret   string    `gorm:"column:app_secret;size:200;not null" json:"app_secret"` // 飞书 App Secret
	Webhook     string    `gorm:"size:500" json:"webhook"`                               // Webhook URL
	Project     string    `gorm:"size:100;not null" json:"project"`                      // 关联项目
	Status      string    `gorm:"size:20;not null" json:"status"`                        // 状态
	Description string    `gorm:"type:text" json:"description"`                          // 描述
	IsDefault   bool      `gorm:"column:is_default;default:false" json:"is_default"`     // 是否默认应用
	CreatedBy   *uint     `gorm:"column:created_by" json:"created_by"`
}

// TableName 指定表名
func (FeishuApp) TableName() string {
	return "feishu_apps"
}

// FeishuRequest 飞书请求存储模型
// 存储飞书交互请求的状态
type FeishuRequest struct {
	gorm.Model
	RequestID       string `gorm:"size:100;not null;uniqueIndex" json:"request_id"` // 请求ID
	OriginalRequest string `gorm:"type:text" json:"original_request"`               // 原始请求 JSON
	DisabledActions string `gorm:"type:text" json:"disabled_actions"`               // 已禁用的操作 JSON
	ActionCounts    string `gorm:"type:text" json:"action_counts"`                  // 操作计数 JSON
}

// TableName 指定表名
func (FeishuRequest) TableName() string {
	return "feishu_requests"
}

// FeishuUserToken 飞书用户令牌模型
// 存储飞书用户 OAuth 令牌
type FeishuUserToken struct {
	ID           uint      `gorm:"primarykey" json:"id"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
	AppID        string    `gorm:"size:100;not null;uniqueIndex" json:"app_id"` // 飞书 App ID
	AccessToken  string    `gorm:"type:text" json:"access_token"`               // 访问令牌
	RefreshToken string    `gorm:"type:text" json:"refresh_token"`              // 刷新令牌
	ExpiresAt    time.Time `json:"expires_at"`                                  // 过期时间
}

// TableName 指定表名
func (FeishuUserToken) TableName() string {
	return "feishu_user_tokens"
}

// FeishuMessageLog 飞书消息发送记录
// 记录所有通过飞书发送的消息
type FeishuMessageLog struct {
	ID            uint      `gorm:"primarykey" json:"id"`
	CreatedAt     time.Time `json:"created_at"`
	MsgType       string    `gorm:"size:50;not null" json:"msg_type"`        // 消息类型: text/post/interactive
	ReceiveID     string    `gorm:"size:100;not null" json:"receive_id"`     // 接收者ID
	ReceiveIDType string    `gorm:"size:50;not null" json:"receive_id_type"` // ID类型: chat_id/open_id/user_id
	Content       string    `gorm:"type:text" json:"content"`                // 消息内容
	Title         string    `gorm:"size:200" json:"title"`                   // 卡片标题
	Source        string    `gorm:"size:50" json:"source"`                   // 来源: manual/oa_sync
	Status        string    `gorm:"size:20;default:'success'" json:"status"` // 状态: success/failed
	ErrorMsg      string    `gorm:"type:text" json:"error_msg"`              // 错误信息
	AppID         uint      `gorm:"column:app_id" json:"app_id"`             // 使用的飞书应用ID
}

// TableName 指定表名
func (FeishuMessageLog) TableName() string {
	return "feishu_message_logs"
}
